android/daemon/src/main/java/org/mozilla/firefox/vpn/daemon/ConnectionHealth.kt

```diff
@@ -0,0 +1,199 @@
+                as ConnectivityManager
+            val vpnNetwork = mVPNNetwork ?: return@Runnable
+            val networkCaps = mConnectivityManager.getNetworkCapabilities(vpnNetwork)
+            if (networkCaps?.hasTransport(NetworkCapabilities.TRANSPORT_VPN) != true) {
+                // In case it's not anymore, just end here.
+                taskDone()
+                return@Runnable
+            }
+
+            // Step 2: Are both the VPN-Gateway and the DNS reachable? (i.e is internet working?)
+            val canReachGateway = vpnNetwork.getByName(gateway).isReachable(PING_TIMEOUT)
+            val canReachDNS = vpnNetwork.getByName(dns).isReachable(PING_TIMEOUT)
+            if (canReachGateway && canReachDNS) {
+                // Internet should be fine :)
+                mResetUsed = false
+                taskDone()
+                return@Runnable
+            }
+
+            if (!canReachGateway)
+                Log.e(TAG, "Failed to Reach VPN-Gateway")
+            if (!canReachDNS)
+                Log.e(TAG, "Failed to reach DNS")
+
+            // Step 3: We have found a connection Issue, check if we can ping the wireguard
+            // endpoint on any Network, maybe the handoff from WIFI-> GSM did not work correctly
+            val anyNetworkCanConnect = mConnectivityManager.allNetworks.firstOrNull() {
+                it.getByName(endpoint).isReachable(PING_TIMEOUT)
+            } != null
+            if (anyNetworkCanConnect && canReachDNS && !mResetUsed) {
+                // The server seems to be online but the connection broke,
+                // Let's just try to force a reconnect ... but only once.
+                mService.mainLooper.run {
+                    // Silent server switch to the same server
+                    mService.reconnect()
+                }
+                mResetUsed = true
+                taskDone()
+                return@Runnable
+            }
+            // Step 4: Either the server is unreachable or the "reconnect" did not
+            // fix the issue. Try to switch to the fallback server.
+            val fallbackServerIsReachable = mConnectivityManager.allNetworks.firstOrNull() {
+                it.getByName(fallbackEndpoint).isReachable(PING_TIMEOUT)
+            } != null
+            if (fallbackServerIsReachable) {
+                Log.i("BASTI", "Switch to fallback VPN server")
+                // We the server is online but the connection broke up, let's rest it
+                mService.mainLooper.run {
+                    // Silent server switch to the same server
+                    mService.reconnect(true)
+                    NotificationUtil.instance?.show(mService)
+                }
+                mResetUsed = true
+                taskDone()
+                return@Runnable
+            }
+            // If we land here: The server and the fallback are unreachable

```

Comment:
If we get this far, can we emit something for the C++ code to handle? For example, we could display the server unavailable notification message or otherwise integrate with the UX via the `Controller::readyToServerUnavailable` signal.